go/parquet/types.go

```diff
@@ -165,3 +169,186 @@ func (fixedLenByteArrayTraits) CastFromBytes(b []byte) []FixedLenByteArray {
+)
+
+// constants for the parquet DataPage Version to use
+const (
+	DataPageV1 DataPageVersion = iota
+	DataPageV2
+)
+
+func (e Encoding) String() string {
+	return format.Encoding(e).String()
+}
+
+var (
+	// Types contains constants for the Physical Types that are used in the Parquet Spec
+	//
+	// They can be specified when needed as such: `parquet.Types.Int32` etc. The values
+	// all correspond to the values in parquet.thrift
+	Types = struct {
+		Boolean           Type
+		Int32             Type
+		Int64             Type
+		Int96             Type
+		Float             Type
+		Double            Type
+		ByteArray         Type
+		FixedLenByteArray Type
+		// this only exists as a convienence so we can denote it when necessary
+		// nearly all functions that take a parquet.Type will error/panic if given
+		// Undefined
+		Undefined Type
+	}{
+		Boolean:           Type(format.Type_BOOLEAN),
+		Int32:             Type(format.Type_INT32),
+		Int64:             Type(format.Type_INT64),
+		Int96:             Type(format.Type_INT96),
+		Float:             Type(format.Type_FLOAT),
+		Double:            Type(format.Type_DOUBLE),
+		ByteArray:         Type(format.Type_BYTE_ARRAY),
+		FixedLenByteArray: Type(format.Type_FIXED_LEN_BYTE_ARRAY),
+		Undefined:         Type(format.Type_FIXED_LEN_BYTE_ARRAY + 1),
+	}
+
+	// Encodings contains constants for the encoding types of the column data
+	//
+	// The values used all correspond to the values in parquet.thrift for the
+	// corresponding encoding type.
+	Encodings = struct {
+		Plain                Encoding
+		PlainDict            Encoding
+		RLE                  Encoding
+		RLEDict              Encoding
+		BitPacked            Encoding // deprecated, not implemented
+		DeltaByteArray       Encoding
+		DeltaBinaryPacked    Encoding
+		DeltaLengthByteArray Encoding
+	}{
+		Plain:                Encoding(format.Encoding_PLAIN),
+		PlainDict:            Encoding(format.Encoding_PLAIN_DICTIONARY),
+		RLE:                  Encoding(format.Encoding_RLE),
+		RLEDict:              Encoding(format.Encoding_RLE_DICTIONARY),
+		BitPacked:            Encoding(format.Encoding_BIT_PACKED),
+		DeltaByteArray:       Encoding(format.Encoding_DELTA_BYTE_ARRAY),
+		DeltaBinaryPacked:    Encoding(format.Encoding_DELTA_BINARY_PACKED),
+		DeltaLengthByteArray: Encoding(format.Encoding_DELTA_LENGTH_BYTE_ARRAY),
+	}
+
+	// ColumnOrders contains constants for the Column Ordering fields
+	ColumnOrders = struct {
+		Undefined        ColumnOrder
+		TypeDefinedOrder ColumnOrder
+	}{
+		Undefined:        format.NewColumnOrder(),
+		TypeDefinedOrder: &format.ColumnOrder{TYPE_ORDER: format.NewTypeDefinedOrder()},
+	}
+
+	// DefaultColumnOrder is to use TypeDefinedOrder
+	DefaultColumnOrder = ColumnOrders.TypeDefinedOrder
+
+	// Repetitions contains the constants for Field Repetition Types
+	Repetitions = struct {
+		Required  Repetition
+		Optional  Repetition
+		Repeated  Repetition
+		Undefined Repetition // convenience value
```

Comment:
it might pay to distinguish beween undefined and not set.  An issue was raised recently that at least in C++ we write Repetition required for the root of the schema when according to the spec we probably shouldn't.

Reply:
in this case `Undefined` == `not set` it explicitly only exists to have an available option for "Not Set". Since "undefined" is not a legitimate value for the parquet spec. If you'd prefer I can change the name of this to be `NotSet` rather than Undefined.